internal/engine/selectors/selectors.go

```diff
@@ -38,8 +39,110 @@ var (
 	ErrResultUnknown = errors.New("result is unknown")
+	// ErrUnsupported is returned when the entity type is not supported
+	ErrUnsupported = errors.New("unsupported entity type")
 )
 
+// ErrKind is a string for the kind of error that occurred
+type ErrKind string
+
+const (
+	// ErrKindParse is an error kind for parsing errors, e.g. syntax errors
+	ErrKindParse ErrKind = "parse"
+	// ErrKindCheck is an error kind for checking errors, e.g. mismatched types
+	ErrKindCheck ErrKind = "check"
+)
+
+// ErrInstance is one occurrence of an error in a CEL expression
+type ErrInstance struct {
+	Line int    `json:"line,omitempty"`
+	Col  int    `json:"col,omitempty"`
+	Msg  string `json:"msg,omitempty"`
+}
+
+// ErrDetails is a struct that holds the details of an error in a CEL expression
+type ErrDetails struct {
+	Errors []ErrInstance `json:"errors,omitempty"`
+	Source string        `json:"source,omitempty"`
+}
+
+// AsJSON returns the ErrDetails as a JSON string
+func (ed *ErrDetails) AsJSON() string {
+	edBytes, err := json.Marshal(ed)
+	if err != nil {
+		return fmt.Sprintf(`{"error": "failed to marshal JSON: %s"}`, err)
+	}
+	return string(edBytes)
+}
+
+func errDetailsFromCelIssues(source string, issues *cel.Issues) ErrDetails {
+	var ed ErrDetails
+
+	ed.Source = source
+	ed.Errors = make([]ErrInstance, 0, len(issues.Errors()))
+	for _, err := range issues.Errors() {
+		ed.Errors = append(ed.Errors, ErrInstance{
+			Line: err.Location.Line(),
+			Col:  err.Location.Column(),
+			Msg:  err.Message,
+		})
+	}
+
+	return ed
+}
+
+// ErrStructure is a struct that callers can use to deserialize the JSON error
+type ErrStructure struct {
+	Err     ErrKind    `json:"err"`
+	Details ErrDetails `json:"details"`
+}
+
+// ParseError is an error type for syntax errors in CEL expressions
+type ParseError struct {
+	ErrDetails
+}
+
+// Error implements the error interface for ParseError
+func (pe *ParseError) Error() string {
+	return fmt.Sprintf(`{"err": "%s", "details": %s}`, ErrKindParse, pe.AsJSON())
+}
+
+// Is checks if the target error is a ParseError
+func (_ *ParseError) Is(target error) bool {
+	var t *ParseError
+	return errors.As(target, &t)
+}
+
+// CheckError is an error type for type checking errors in CEL expressions, e.g.
+// mismatched types
+type CheckError struct {
+	ErrDetails
+}
+
+// Error implements the error interface for CheckError
+func (ce *CheckError) Error() string {
+	return fmt.Sprintf(`{"err": "%s", "details": %s}`, ErrKindCheck, ce.AsJSON())
+}
+
+// Is checks if the target error is a CheckError
+func (_ *CheckError) Is(target error) bool {
+	var t *CheckError
+	ok := errors.As(target, &t)
+	return ok
+}
+
+func newParseError(source string, issues *cel.Issues) error {
+	return &ParseError{
+		ErrDetails: errDetailsFromCelIssues(source, issues),
+	}
+}
+
+func newCheckError(source string, issues *cel.Issues) error {
+	return &CheckError{
+		ErrDetails: errDetailsFromCelIssues(source, issues),
+	}
+}
+
```

Comment:
should these errors wrap a common one to ease error handling? https://pkg.go.dev/errors you could implement `Unwraps` for this.

Reply:
that's a really good idea!

Reply:
done. We could even use `errors.Join(ErrCheckFailed, ErrParse)` for parse and join with a check error for the check but I'm not sure if it's worth it.